### PR TITLE
etcd dashboard with variable to select datasource

### PR DESCRIPTION
**Pull request by iblackman**
The current dashboard provisioned by prometheus with ETCD metrics doesn't support our single grafana with multiple datasources. This dashboard is basically a copy of it with the datasource variable being used in all panels
Screen Shot 2022-02-01 at 7 28 07 PM
.

### CODE REVIEW

File: etcd-by-prometheus-with-datasource.json [hunk 1]
```diff
@@ -0,0 +1,1857 @@
...
+        "dashes": false,
+        "datasource": "$datasource",
+        "description": "proposals_committed_total records the total number of consensus proposals committed. This gauge should increase over time if the cluster is healthy. Several healthy members of an etcd cluster may have different total committed proposals at once. This discrepancy may be due to recovering from peers after starting, lagging behind the leader, or being the leader and therefore having the most commits. It is important to monitor this metric across all the members in the cluster; a consistently large lag between a single member and its leader indicates that member is slow or unhealthy.\n\nproposals_applied_total records the total number of consensus proposals applied. The etcd server applies every committed proposal asynchronously. The difference between proposals_committed_total and proposals_applied_total should usually be small (within a few thousands even under high load). If the difference between them continues to rise, it indicates that the etcd server is overloaded. This might happen when applying expensive queries like heavy range queries or large txn operations.",
```
Comment (jackiehpowerhrg) [inline etcd-by-prometheus-with-datasource.json:1360]: Just a few questions for my understanding:
1. Does it matter which node is the leader at any point in time? Or does it matter more that the cluster has a leader regardless of which one it is?
2. What happens when there's no leader? What is the leader responsible for? I found documentation that says 'Etcd is a crucial component for Kubernetes as it stores the entire state of the cluster: its configuration, specifications, and the statuses of the running workloads.', so the leader is the main reference for how kube is configured?

File: etcd-by-prometheus-with-datasource.json [hunk 2]
```diff
@@ -0,0 +1,1857 @@
...
+          }
+        ]
+      },
+      {
+        "aliasColors": {},
+        "bars": false,
+        "dashLength": 10,
+        "dashes": false,
+        "datasource": "$datasource",
+        "description": "indicates how many proposals are queued to commit. Rising pending proposals suggests there is a high client load or the member cannot commit proposals.",
```
Comment (jackiehpowerhrg) [inline etcd-by-prometheus-with-datasource.json:1455]: Proposals are basically requests correct? Is there a threshold you expect each member to have queued at a given time? If so how did you get to that number?

File: etcd-by-prometheus-with-datasource.json [hunk 3]
```diff
@@ -0,0 +1,1857 @@
...
+        "seriesOverrides": [],
+        "spaceLength": 10,
+        "stack": false,
+        "steppedLine": false,
+        "targets": [
+          {
+            "expr": "sum(rate(etcd_disk_wal_fsync_duration_seconds_sum[1m]))",
+            "format": "time_series",
+            "intervalFactor": 2,
+            "legendFormat": "\tThe latency distributions of fsync called by wal",
```
Comment (jackiehpowerhrg) [inline etcd-by-prometheus-with-datasource.json:1573]: Is this supposed to have the '\t' in front of it?

**Comment by jackiehpowerhrg** [comment]
@bengoa I wonder if we can do something similar to this for the core-infra metrics. I think our dashboards are manually configured within the GUI right?

**Comment by iblackman** [comment]
> > @bengoa I wonder if we can do something similar to this for the core-infra metrics. I think our dashboards are manually configured within the GUI right?
> 
> yes, we definitively can. But what I believe Watchtower is doing is creating the dashboard at the GUI and then exporting the JSON for git versioning purposes. This particular one (etcd) may have been made by the same guys who have provided the chart (or something like this).
> 
> Creating a Dashboard from scratch in JSON is burdensome.
> 
> @iblackman thoughts?

Exaclty, I don't create in JSON haha I alter/create in the GUI and then save the JSON model to this repo. That dashboard itself is already provided as a default dashboard from prometheus, but it doesn't have the datasource as variable.

File: etcd-by-prometheus-with-datasource.json (hunk 1 above)
Comment (iblackman) [reply etcd-by-prometheus-with-datasource.json:1360]: 1. it doesn't matter who is the leader. The cluster can't work without a leader.
2. Think about the leader as the active, and the other members as backup. The leader is the active node of the kube cluster "database". ETCD is kubernetes' database.

File: etcd-by-prometheus-with-datasource.json (hunk 2 above)
Comment (iblackman) [reply etcd-by-prometheus-with-datasource.json:1455]: Proposals are the numbers of changes in the data. Think as a version almost. Every new/updated data received by the leader makes that number to increase and then the leader replicate to the other members which change their number too.

File: etcd-by-prometheus-with-datasource.json (hunk 3 above)
Comment (iblackman) [reply etcd-by-prometheus-with-datasource.json:1573]: I didn't touch in anything of this chart besides the `"datasource": "$datasource",`. All the rest comes from the prometheus repository itself.

File: etcd-by-prometheus-with-datasource.json (hunk 2 above)
Comment (jackiehpowerhrg) [reply etcd-by-prometheus-with-datasource.json:1455]: Is this done so if something goes wrong with the data there's a proposal (version) to go back to?

**Review by jackiehpowerhrg — APPROVED** [review]

File: etcd-by-prometheus-with-datasource.json (hunk 2 above)
Comment (iblackman) [reply etcd-by-prometheus-with-datasource.json:1455]: No idea

**Pull request merged**